src/cunumeric/fft/fft.cu

```diff
@@ -0,0 +1,351 @@
+/* Copyright 2021 NVIDIA Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#include <csignal>
+
+#include "cunumeric/fft/fft.h"
+#include "cunumeric/fft/fft_template.inl"
+
+#include "cunumeric/cuda_help.h"
+
+namespace cunumeric {
+
+using namespace Legion;
+using namespace legate;
+
+template <int DIM>
+__host__ static inline void cufft_operation(void* output,
+                                            void* input,
+                                            const Rect<DIM>& out_rect,
+                                            const Rect<DIM>& in_rect,
+                                            std::vector<int64_t>& axes,
+                                            fftType type,
+                                            fftDirection direction)
+{
+    cudaStream_t stream;
+    CHECK_CUDA(cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking));
+
+    size_t workarea_size;
+    size_t num_elements;
+    int n[DIM];
+    int inembed[DIM];
+    int onembed[DIM];
+
+    const Point<DIM> zero   = Point<DIM>::ZEROES();
+    const Point<DIM> one    = Point<DIM>::ONES();
+    Point<DIM> fft_size_in  =  in_rect.hi -  in_rect.lo + one;
+    Point<DIM> fft_size_out = out_rect.hi - out_rect.lo + one;
+    num_elements = 1;
+    for(int i = 0; i < DIM; ++i) {
+      n[i]          = (type == fftType::FFT_R2C || type == fftType::FFT_D2Z) ? fft_size_in[i] : fft_size_out[i];
+      inembed[i]    = fft_size_in[i];
+      onembed[i]    = fft_size_out[i];
+      num_elements *= n[i];
+    }
+
+    // Create the plan
+    cufftHandle plan;
+    CHECK_CUFFT(cufftCreate(&plan));
+    CHECK_CUFFT(cufftSetAutoAllocation(plan, 0 /*we'll do the allocation*/));
+    CHECK_CUFFT(cufftSetStream(plan, stream));
+
+    // Create the plan and allocate a temporary buffer for it if it needs one
+    CHECK_CUFFT(cufftMakePlanMany(plan, DIM, n, inembed, 1, 1, onembed, 1, 1, (cufftType)type, 1, &workarea_size));
+
+    DeferredBuffer<uint8_t, 1> workarea_buffer;
+    if(workarea_size > 0) {
+      const Point<1> zero1d(0);
+      workarea_buffer =
+        DeferredBuffer<uint8_t, 1>(Rect<1>(zero1d, Point<1>(workarea_size - 1)),
+                                   Memory::GPU_FB_MEM,
+                                   nullptr /*initial*/,
+                                   128 /*alignment*/);
+      void* workarea = workarea_buffer.ptr(zero1d);
```

Comment:
Is the buffer created (or fetched if one already exist) at this point? Is `workarea_buffer.ptr` blocking until a buffer becomes available?